### PR TITLE
build(node): upgrade to node v20.11.0 and related dependency

### PR DESCRIPTION
**Pull request by rockleona**
## Types of changes

- [x] **Other (please describe)**

## Description

- Update Node to v20.11.0 (lts/iron)
- Update some packages
  - autoprefixer
  - browserify
  - node-sass
  - postcss-cli

## Related Issue

Ideal to resolve all [Dependabot alerts](https://github.com/pycontw/pycon.tw/security/dependabot)

### CODE REVIEW

**Comment by rockleona** [comment]
I was trying to run dev container on Windows WSL2 (Ubuntu 22.04), but will got this error when running `docker-compose-dev.yml`.

Will try to use Ubuntu tomorrow.

image

**Comment by mattwang44** [comment]
yap I've aware that the script cannot run on Windows...still WIP
https://github.com/pycontw/pycon.tw/pull/1192

**Comment by mattwang44** [comment]
should also edit the version defined in `.github/workflows/ci.yaml` for CI

**Comment by rockleona** [comment]
> hmmm it should have worked on WSL. Which docker version do you use now?

Wondering if this info is helpful or not

image

**Comment by rockleona** [comment]
> should also edit the version defined in `.github/workflows/ci.yaml` for CI

Sure, I just notice the ci error too, I think I will wait #1194 merged and edit all node version at once.

**Comment by rockleona** [comment]
Hi @mattwang44, things getting weird now..., I checkout master branch to run `enter_dev_env.sh` on Ubuntu22.04, but still having this error, I'm wondering how should I do  in this stage.😢 

> Error
image

> Docker Version
image

**Comment by mattwang44** [comment]
QQ Would you please try to run the script on GitHub Codespace (it was functional a couple of weeks ago)? 
I'm afraid that I do not have enough time to cope with this issue until next week.

**Comment by rockleona** [comment]
Hi @mattwang44 , I think I'm making the first step done upgrading the node packages, please have a review.

**Comment by mattwang44** [comment]
有 conflict 記得先解掉歐～

File: Dockerfile [hunk 1]
```diff
@@ -1,21 +1,25 @@
...
 FROM python_base as python_deps
 ENV APP_DIR /usr/local/app
 
+# make nodejs accessible and executable globally
+ENV NODE_PATH $APP_DIR/node_modules/
+ENV PATH /usr/local/bin:$PATH
+
+# Add bin directory used by `pip install --user`
+ENV PATH /home/docker/.local/bin:$PATH
+
```
Comment (rockleona) [inline Dockerfile:22]: 這邊是不是不需要了＠＠

File: src/pycontw2016/settings/base.py [hunk 2]
```diff
@@ -265,7 +265,7 @@ def node_bin(name):
 COMPRESS_SCSS_COMPILER_CMD = (
     '{node_sass_bin} --include --output-style expanded {paths} '
     '--include-path="{node_modules}" "{infile}" "{outfile}" && '
-    '{postcss_bin} --use "{node_modules}/autoprefixer" '
+    '{postcss_bin} --use "{node_modules}/autoprefixer/lib/autoprefixer.js" '
```
Comment (rockleona) [inline src/pycontw2016/settings/base.py:268]: [原本文件中的command](https://github.com/kottenator/django-compressor-toolkit?tab=readme-ov-file#compress_scss_compiler_cmd) 會抓不到其他套件，所以改用這樣的方式處理

**Review by rockleona — COMMENTED** [review]
我 rebase 了！

File: dev.Dockerfile
Comment (mattwang44) [inline dev.Dockerfile:1]: #1194 後沒有這份檔案囉

File: Dockerfile (hunk 1 above)
Comment (mattwang44) [reply Dockerfile:22]: 拿掉&重 build 後 frontend 會有問題嗎？

Comment (rockleona) [reply Dockerfile:22]: 看起來沒問題，那我就拿掉了喔

File: yarn.lock
Comment (mattwang44) [inline yarn.lock:1]: 這裡有 conflict 被 commit，要記得重 generate
image

File: Dockerfile [hunk 3]
```diff
@@ -1,17 +1,14 @@
...
 COPY ./yarn.lock yarn.lock
 COPY ./package.json package.json
 
-RUN apt-get update
-RUN apt-get install python-pip -y
-
-RUN npm install -g yarn
-RUN yarn install --dev --frozen-lockfile && yarn cache clean
+RUN corepack enable
+RUN yarn install --immutable
```
Comment (mattwang44) [inline Dockerfile:11]: 這邊指令的改變對 image build 有什麼影響或好處嗎？

File: package.json [hunk 4]
```diff
@@ -12,11 +12,12 @@
...
     "@babel/preset-env": "^7.23.2",
-    "autoprefixer": "^8.0.0",
+    "autoprefixer": "^10.4.19",
     "babelify": "^10.0.0",
-    "browserify": "^16.1.0",
-    "node-sass": "^4.9.0",
-    "postcss-cli": "^5.0.0"
+    "browserify": "^17.0.0",
+    "node-sass": "^9.0.0",
+    "postcss": "^8.4.38",
```
Comment (mattwang44) [inline package.json:19]: 多了 "postcss"，不加會有問題嗎？

File: Dockerfile (hunk 3 above)
Comment (rockleona) [reply Dockerfile:11]: 原先的安裝方法是 yarn1 的[安裝方法](https://classic.yarnpkg.com/en/docs/install#debian-stable)，
因為現在升級 node 到 v20 且用 latest yarn，[啟用方法](https://yarnpkg.com/getting-started/install)已經改變

File: package.json (hunk 4 above)
Comment (rockleona) [reply package.json:19]: 不知道為什麼 `postcss-cli`@11.x 會抓不到 `postcss`，
安裝之後才找得到這個套件。

File: yarn.lock
Comment (rockleona) [reply yarn.lock:1]: > 只差這個，改完就 approve & merge~

剛推上來喔~

**Review by mattwang44 — APPROVED** [review]

**Comment by mattwang44** [comment]
@SivanYeh please help to merge/deploy this PR whenever you think it's good to go

**Comment by SivanYeh** [comment]
> @SivanYeh please help to merge/deploy this PR whenever you think it's good to go
There're several collaborations going on the same platforms within two weeks. 
I may act more conservative atm lol. 
Let me deploy a test branch "temp_master" these two weeks. If everything goes smooth, I will merge then.

**Review by SivanYeh — APPROVED** [review]
LGTM! It worked well during these two weeks. I will merge it into master.
Thanks all!

**Pull request merged**